transport/transporters/kafka/transporter/transporter.go

```diff
@@ -0,0 +1,216 @@
+package transporter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Nextdoor/pg-bifrost.git/marshaller"
+	"github.com/Nextdoor/pg-bifrost.git/shutdown"
+	"github.com/Nextdoor/pg-bifrost.git/stats"
+	"github.com/Nextdoor/pg-bifrost.git/transport"
+	"github.com/Nextdoor/pg-bifrost.git/transport/batch"
+	"github.com/Nextdoor/pg-bifrost.git/utils"
+	"github.com/Shopify/sarama"
+	"github.com/cenkalti/backoff/v4"
+	"github.com/cevaris/ordered_map"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	TimeSource utils.TimeSource = &utils.RealTime{}
+)
+
+type KafkaTransporter struct {
+	shutdownHandler shutdown.ShutdownHandler
+	inputChan       <-chan transport.Batch         // receive a batch (slice) of MarshalledMessages
+	txnsWritten     chan<- *ordered_map.OrderedMap // Map of <transaction:progress.Written>
+	statsChan       chan stats.Stat
+	log             logrus.Entry
+	kafkaProducer   sarama.SyncProducer
+	retryPolicy     backoff.BackOff
+	topic           string
+}
+
+func NewTransporter(
+	shutdownHandler shutdown.ShutdownHandler,
+	inputChan <-chan transport.Batch,
+	statsChan chan stats.Stat,
+	txnsWritten chan<- *ordered_map.OrderedMap,
+	log logrus.Entry,
+	retryPolicy backoff.BackOff,
+	producer sarama.SyncProducer,
+	topic string) transport.Transporter {
+
+	return &KafkaTransporter{
+		shutdownHandler,
+		inputChan,
+		txnsWritten,
+		statsChan,
+		log,
+		producer,
+		retryPolicy,
+		topic,
+	}
+
+}
+
+// shutdown idempotently closes the output channel
+func (t *KafkaTransporter) shutdown() {
+	t.log.Info("shutting down transporter")
+	_ = t.kafkaProducer.Close()
+	t.shutdownHandler.CancelFunc() // initiate shutdown on other modules as well
+
+	if r := recover(); r != nil {
+		t.log.Warnf("Recovered in KafkaTransporter %s", r)
+	}
+
+	defer func() {
+		// recover if channel is already closed
+		_ = recover()
+	}()
+
+	t.log.Debug("closing progress channel")
+	close(t.txnsWritten)
+}
+
+func (t *KafkaTransporter) transportWithRetry(ctx context.Context, produceMessages []*sarama.ProducerMessage) (error, bool) {
+	var cancelled bool
+	var ts = TimeSource
+
+	operation := func() error {
+		select {
+		case <-ctx.Done():
+			t.log.Debug("received terminateCtx cancellation")
+			cancelled = true
+			return nil
+		default:
+		}
+
+		err := t.kafkaProducer.SendMessages(produceMessages)
+		if err == nil {
+			t.statsChan <- stats.NewStatCount("kafka_transport", "success", 1, ts.UnixNano())
+			return nil
+		}
+
+		produceErrors, ok := err.(sarama.ProducerErrors)
+		if !ok {
+			panic("produceErrors is not type sarama.ProducerErrors")
```

Comment:
What are cases where the err isn't a `sarama.ProducerErrors`?

Reply:
The `SendMessages` function signature uses the `error` type for it's return value, so the returned `err` value from `SendMessages` will need to be type casted. However, the only actual return type of the `SendMessages` fn other than nil is `sarama.ProducerErrors`, so this panic check isn't necessary.